### PR TITLE
E2E test for FSDP, HSDP, FSDP+TP in Distributed Checkpointing

### PR DESCRIPTION
**Pull request by LucasLLC**
Adds E2E tests for saving/loading distributed checkpoints. Supported so far are:

- FSDP
- HSDP
- FSDP + TP

Each method is also tested using `torch.compile`

To run all tests:
`python test/distributed/checkpoint/test/distributed/checkpoint/e2e/test_e2e_save_and_load.py`

### CODE REVIEW

File: test/distributed/checkpoint/e2e/test_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        dummy_model = TestDummyModel()
+
+        model = FSDP(dummy_model.cuda(), device_mesh=device_mesh, use_orig_params=True)
+        optim = torch.optim.Adam(model.parameters(), lr=0.1)
+        model(model.get_input()).sum().backward()
+        optim.step()
+
+        return model, optim
+
+    def _equal_state_dict(self, model_0, model_1):
```
Comment (LucasLLC) [inline test/distributed/checkpoint/e2e/test_fsdp.py:51]: @fegin Is your recommendation (https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_fsdp_2d_parallel.py#L638-L641) necessary here, or do you think this is sufficient?

Comment (LucasLLC) [reply test/distributed/checkpoint/e2e/test_fsdp.py:51]: Assuming we'll need something similar to compare optimizers?

Comment (wz337) [reply test/distributed/checkpoint/e2e/test_fsdp.py:51]: Chien-Chin has some utils to compare state_dict here. See if anything is useful for you, and maybe we can just create a test utils for these. https://github.com/pytorch/pytorch/blob/main/test/distributed/checkpoint/test_state_dict.py#L56

Comment (LucasLLC) [reply test/distributed/checkpoint/e2e/test_fsdp.py]: I believe since I can compare DTensor's directly with `torch.equal` we don't really need this logic, so I left the model state dict comparison the same for now.

I also opted for a direct comparison of the optim state dict, since it was a bit simpler and that seems to work here as well - https://github.com/pytorch/pytorch/blob/5a6f8014c4c2480f77d40fa214ca43705d90f7a1/test/distributed/checkpoint/test_state_dict.py#L146 .

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        return model, optim
+
+    def _equal_state_dict(self, model_0, model_1):
+        for params_0, params_1 in zip(model_0.values(), model_1.values()):
+            if not torch.equal(params_0, params_1):
+                return False
+        return True
+
+    @with_comms
+    @skip_if_lt_x_gpu(2)
```
Comment (wz337) [inline test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:96]: I don't think FSDP + TP work for GPU = 2. Your local test is probably fine, since we have 8 GPUs on DEVGPU. We will need @skip_if_lt_x_gpu(4).

**Comment by wz337** [comment]
Adding a few tags. @LucasLLC Just FYI, the multi-gpu tests do not run by default. To enable the multi gpu tests, we need to add CI/periodic to the PR.

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+class ModelType(Enum):
```
Comment (wz337) [inline test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:44]: Very clean!!

**Review by H-Huang — COMMENTED** [review]
Looks great! I would also include an example python / pytest command in the PR summary to include as a reference.

Will defer the approval stamp to the DCP experts :)

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    @with_comms
+    @skip_if_lt_x_gpu(2)
+    @with_temp_dir
+    @parametrize("compile", [True, False])
+    @parametrize("model_type", [ModelType.FSDP, ModelType.HSDP, ModelType.FSDP_TP])
+    def test_e2e(self, compile, model_type):
+        # first create and save a checkpoint
+        model, optim = self._create_model(compile, model_type)
+        model_state_dict_0, optim_state_dict_0 = get_state_dict(model, optimizers=optim)
```
Comment (fegin) [inline test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:103]: I suggest that we use a non-parallelized model and directly call state_dict as the source of truth to compare.

1. create a non-parallelized model
2. create a parallelized model
3. train both models 2 steps
4. save the parallelized model
5. create a new parallelized model
6. load from the trained parallelized model to the new parallelized model
7. train another step both the new parallelized model and the non-parallelized model and compare the accuracy.

**Review by fegin — APPROVED** [review]
LGTM, we can land this first and improve it to compare the result with a non-parallelized model.

Please fix the lint issue and skip_if_lt_x_gpu issue @wz337 mentioned before landing.

File: test/distributed/checkpoint/e2e/test_fsdp.py (hunk 1 above)
Comment (fegin) [reply test/distributed/checkpoint/e2e/test_fsdp.py]: If you are comparing 2 identical state_dicts, this should be enough. However, if you need to compare with non-converted optimizer state_dict, a customized function is required as non-converted optimizer state use parameter id.

**Review by wz337 — APPROVED** [review]
Thanks for having this ready! Super-duper fast! 
With this PR, now we can mark our FSDP, HSDP, 2D checkpointing B/E ready on PT-D feature support matrix! Thanks @LucasLLC!

**Comment by fegin** [comment]
Try `lintrunner` locally to get the signal of lint and typing errors.

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                use_orig_params=True,
+                sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+            )
+        elif model_type == ModelType.FSDP_TP:
+            mesh_2d = init_device_mesh(
+                self.device_type, (2, self.world_size // 2), mesh_dim_names=("dp", "tp")
+            )
+            tp_mesh = mesh_2d["tp"]
+            dp_mesh = mesh_2d["dp"]
+            model = parallelize_module(dummy_model, tp_mesh, PairwiseParallel())
```
Comment (wanchaol) [inline test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:76]: As a follow up, we should probably switch `PairwiseParallel` to `Colwise + Rowwise` as we plan to deprecate the former soon

Comment (wz337) [reply test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:76]: @LucasLLC 

For usage of Colwise + Rowwise, you can refer to this slide. https://docs.google.com/presentation/d/1e9TNYu_u_Hz9IpfhS4_R_77de59g5AnSWEf7U-FDphg/edit#slide=id.g28cb00a468a_0_16 

So, for this model, we can specify a parallelize_plan for some of the layers, maybe layer 1 and layer 2. https://docs.google.com/presentation/d/1e9TNYu_u_Hz9IpfhS4_R_77de59g5AnSWEf7U-FDphg/edit#slide=id.g28cb00a468a_0_16

Comment (fduwjj) [reply test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:76]: actually since we use sequential in the module definition, I am not sure if `PairwiseParallel` is working here. So using Colwise + Rowwise might be a necessary. Basically, you need to pass in a dictionary as a plan to `parallelize_module` with its key equals to module's FQN and value is ColwiseParallel or RowwiseParallel.
